opengever/maintenance/scripts/make_repository_readonly.py

```diff
@@ -0,0 +1,363 @@
+            raise Exception("Object %r is not a repository root!" % self.reporoot)
+
+    def run(self):
+        self.assert_no_open_dossiers(self.reporoot)
+        self.assert_no_open_tasks(self.reporoot)
+
+        with VerifyViewPermissionUnchanged(self.reporoot, enabled=self.verify):
+            with OperationsLog(self.repository_id) as ops_log:
+                self.drop_all_roles_except_reader(self.reporoot, ops_log)
+
+        logger.info("Successfully made repository %r readonly." % self.repository_id)
+        if self.dryrun:
+            logger.info("(Dry Run - no changes committed)")
+
+    def drop_all_roles_except_reader(self, reporoot, ops_log):
+        for obj in RepositoryContentIterator(self.reporoot, yield_root=True):
+            logger.info("Checking " + get_path(obj))
+
+            self.assert_in_closed_state(obj)
+
+            manager = RoleAssignmentManager(obj)
+            for assignment_data in manager.storage.get_all():
+                old_assignment = dict(assignment_data)
+                new_assignment = dict(assignment_data)
+                new_assignment["roles"] = [u"Reader"]
+
+                self.assert_no_unexpected_roles_dropped(old_assignment["roles"])
+
+                manager.add_or_update(
+                    principal=new_assignment["principal"],
+                    roles=new_assignment["roles"],
+                    cause=new_assignment["cause"],
+                    reference=new_assignment["reference"],
+                    reindex=False,
+                )
+                ops_log.local_roles_modified(obj, old_assignment, new_assignment)
+
+    def assert_no_open_dossiers(self, reporoot):
+        """Quick check using catalog to make sure no open dossiers exist.
+        """
+        self._assert_no_open_objects(
+            reporoot, "dossiers", IDossierMarker, DOSSIER_STATES_OPEN
+        )
+
+    def assert_no_open_tasks(self, reporoot):
+        """Quick check using catalog to make sure no open tasks exist.
+        """
+        self._assert_no_open_objects(
+            reporoot, "tasks", ITask, OPEN_TASK_STATES,
+        )
+
+    def _assert_no_open_objects(self, reporoot, type_name, type_iface, open_states):
+        catalog = api.portal.get_tool("portal_catalog")
+        open_objects = catalog.unrestrictedSearchResults(
+            path=get_path(reporoot),
+            object_provides=type_iface.__identifier__,
+            review_state=open_states,
+        )
+        if len(open_objects) > 0:
+            logger.error("The following %s are not closed:" % type_name)
+            for brain in open_objects:
+                logger.error(brain.getPath())
+            raise Exception("Found open %s!" % type_name)
+
+    def assert_in_closed_state(self, obj):
+        """Check again that dossiers and tasks are closed (using the obj).
+        """
+        if IDossierMarker.providedBy(obj):
+            if api.content.get_state(obj) not in DOSSIER_STATES_CLOSED:
```

Comment:
dossier-state-offered is neither in the `DOSSIER_STATES_CLOSED` nor in the `DOSSIER_STATES_OPEN`